bottlecap/src/traces/span_pointers.rs

```diff
@@ -0,0 +1,198 @@
+// Copyright 2023-Present Datadog, Inc. https://www.datadoghq.com/
+// SPDX-License-Identifier: Apache-2.0
+
+use datadog_trace_protobuf::pb::SpanLink;
+use sha2::{Digest, Sha256};
+use std::collections::HashMap;
+
+const SPAN_POINTER_HASH_LENGTH: usize = 32;
+
+#[derive(Clone)]
+pub struct SpanPointer {
+    pub hash: String,
+    pub kind: String,
+}
+
+/// Returns the first 32 characters of the SHA-256 hash of the components joined by a '|'.
+/// Used by span pointers to uniquely & deterministically identify an `S3` or `DynamoDB` stream.
+/// <https://github.com/DataDog/dd-span-pointer-rules/blob/main/README.md#General%20Hashing%20Rules>
+#[must_use]
+pub fn generate_span_pointer_hash(components: &[&str]) -> String {
+    let mut hasher = Sha256::new();
+    hasher.update(components.join("|").as_bytes());
+    let result = hasher.finalize();
+    hex::encode(result)[..SPAN_POINTER_HASH_LENGTH].to_string()
+}
+
+pub fn attach_span_pointers_to_meta<S: ::std::hash::BuildHasher>(
+    meta: &mut HashMap<String, String, S>,
+    span_pointers_option: &Option<Vec<SpanPointer>>,
+) {
+    let Some(span_pointers) = span_pointers_option else {
+        return;
+    };
+
+    if span_pointers.is_empty() {
+        return;
+    }
+
+    let span_links: Vec<SpanLink> = span_pointers
+        .iter()
+        .map(|sp| {
+            let mut attributes = HashMap::new();
+            attributes.insert("link.kind".to_string(), "span-pointer".to_string());
+            attributes.insert("ptr.dir".to_string(), "u".to_string());
+            attributes.insert("ptr.hash".to_string(), sp.hash.clone());
+            attributes.insert("ptr.kind".to_string(), sp.kind.clone());
+
+            SpanLink {
+                // We set trace_id and span_id to 0 since they're unknown; the frontend
+                // uses `ptr.hash` instead to find the opposite link if it exists.
+                trace_id: 0,
+                span_id: 0,
+                attributes,
```

Comment:
Move the attributes to the top/bottom, and then explain that not only trace_id or span_id, but all trace propagation fields